bgpd/bgp_nexthop.c

```diff
@@ -806,97 +949,191 @@ static void bgp_show_nexthops_detail(struct vty *vty, struct bgp *bgp,
+			if (CHECK_FLAG(bnc->flags, BGP_NEXTHOP_CONNECTED))
+				vty_out(vty, "  Must be Connected\n");
+			if (!CHECK_FLAG(bnc->flags, BGP_NEXTHOP_REGISTERED))
+				vty_out(vty, "  Is not Registered\n");
+		}
 	}
 	tbuf = time(NULL) - (monotime(NULL) - bnc->last_update);
-	vty_out(vty, "  Last update: %s", ctime(&tbuf));
+	if (json) {
+		if (!specific) {
+			json_last_update = json_object_new_object();
+			json_object_int_add(json_last_update, "epoch", tbuf);
+			json_object_string_add(json_last_update, "string",
+					       ctime(&tbuf));
+			json_object_object_add(json_nexthop, "lastUpdate",
+					       json_last_update);
+		} else {
+			json_object_int_add(json_nexthop, "lastUpdate", tbuf);
+		}
+	} else {
+		vty_out(vty, "  Last update: %s", ctime(&tbuf));
+	}
 
 	/* show paths dependent on nexthop, if needed. */
 	if (specific)
-		bgp_show_nexthop_paths(vty, bgp, bnc);
+		bgp_show_nexthop_paths(vty, bgp, bnc, json_nexthop);
+	if (json)
+		json_object_object_add(json, buf, json_nexthop);
 }
 
 static void bgp_show_nexthops(struct vty *vty, struct bgp *bgp,
-			      bool import_table)
+			      bool import_table, json_object *json, afi_t afi,
+			      bool detail)
 {
 	struct bgp_nexthop_cache *bnc;
-	afi_t afi;
 	struct bgp_nexthop_cache_head(*tree)[AFI_MAX];
+	json_object *json_afi = NULL;
+	bool found = false;
 
-	if (import_table)
-		vty_out(vty, "Current BGP import check cache:\n");
-	else
-		vty_out(vty, "Current BGP nexthop cache:\n");
+	if (!json) {
+		if (import_table)
+			vty_out(vty, "Current BGP import check cache:\n");
+		else
+			vty_out(vty, "Current BGP nexthop cache:\n");
+	}
 	if (import_table)
 		tree = &bgp->import_check_table;
 	else
 		tree = &bgp->nexthop_cache_table;
+
+	if (afi == AFI_IP || afi == AFI_IP6) {
+		if (json)
+			json_afi = json_object_new_object();
+		frr_each (bgp_nexthop_cache, &(*tree)[afi], bnc) {
+			bgp_show_nexthop(vty, bgp, bnc, detail, json_afi);
+			found = true;
+		}
+		if (found && json)
+			json_object_object_add(
+				json, (afi == AFI_IP) ? "ipv4" : "ipv6",
+				json_afi);
+
+		return;
+	}
+
 	for (afi = AFI_IP; afi < AFI_MAX; afi++) {
+		if (json && (afi == AFI_IP || afi == AFI_IP6))
+			json_afi = json_object_new_object();
 		frr_each (bgp_nexthop_cache, &(*tree)[afi], bnc)
-			bgp_show_nexthop(vty, bgp, bnc, false);
+			bgp_show_nexthop(vty, bgp, bnc, detail, json_afi);
+		if (json && (afi == AFI_IP || afi == AFI_IP6))
+			json_object_object_add(
+				json, (afi == AFI_IP) ? "ipv4" : "ipv6",
+				json_afi);
 	}
 }
 
 static int show_ip_bgp_nexthop_table(struct vty *vty, const char *name,
-				     const char *nhopip_str,
-				     bool import_table)
+				     const char *nhopip_str, bool import_table,
+				     json_object *json, afi_t afi, bool detail)
 {
 	struct bgp *bgp;
 
-	if (name)
+	if (name && strcmp(name, VRF_DEFAULT_NAME) != 0)
 		bgp = bgp_lookup_by_name(name);
 	else
 		bgp = bgp_get_default();
 	if (!bgp) {
-		vty_out(vty, "%% No such BGP instance exist\n");
+		if (!json)
```

Comment:
Shouldn't we return an empty `{}` JSON as well for JSON output?

Reply:
show_ip_bgp_nexthop_table() is called with an empty JSON object ({}). So, when show_ip_bgp_nexthop_table() returns CMD_WARNING, empty JSON object will be printed by the caller of show_ip_bgp_nexthop_table().

Reply:
Here are the outputs:

r2# show bgp vrf fed nexthop 
% No such BGP instance exist
r2# show bgp vrf fed nexthop json 
{
}